### ISSUE TITLE
CJ4: Legs and Progress page random (dis)appearance of WPTs

### ISSUE THREAD

**Issue by TheFrett** [Working-Title-MSFS-Mods/fspackages#220]
This one may be a bit difficult to reproduce. I will add more detail to this issue if I come across this more often. On final approach @EGNM in the below video, not all my final approach waypoints show up in the LEGS page. Going to the PROGRESS page does show them correctly. When I go back to the LEGS page, a past waypoint (DME09) suddenly appends to the already displayed legs. Back in the PROGRESS page now everything is gone. Hard to follow I suppose, see what I mean in the video below:

Version used is 0.4.0 release

https://youtu.be/fwA_erS_17M

**Comment by dga711** [comment]
Afaik the PROG page didn't receive a lot of work yet and is mostly work in progress.

Nonetheless, that waypoints don't show up on LEGS and suddenly appear shouldn't happen. 
Will have to look into what is going on there as in general, we just grab the waypoints the game is giving to us. 

Could you write down your flight plan and approach, so we can test it?

Thank you for your feedback with all the details and even videos ;)

**Comment by TheFrett** [comment]
EHBK/21 OSGOS1B OSGOS LEKKO GOLES RIMTO EGNM/32 (ILS32 approach via LBA)

Flightplan was entered manually after completing a short leg from EDDL to EHBK 

As I said, it may be hard to reproduce. I was doing some waypoint switching and inserting (having fun with the new features :D) so I will report if I come across this in more controlled test conditions.

**Comment by dga711** [comment]
Thats nice.
We also noticed during testing that we sometimes hit weird states of the flight plan when playing with, especially when nearing approach. So every example helps.

**Comment by TheFrett** [comment]
What I also noticed, but I don't have a screenshot or video of this, is that at the time this happened the FMS TEXT display in the upper MFD menu also got a mind of its own, showing the departure airport as next waypoint with ~350 nm to go and then following the original flight plan back to EGNM again (so the final waypoint appeared as ~700 nm to go). Basically there were three versions of the active flight plan displayed with the PROGRESS, LEGS pages of the FMS and the FMS TEXT on the MFD each having a mind of their own.

Hopefully I can provide more examples to help debugging...

**Comment by TheFrett** [comment]
Just did a flight: 

ESSK/18 IBGAX EBURI AAL EKYT/26R (ILS approach transition AAL), all manually entered on the ramp.

To illustrate the inconsistencies:

The FMS TEXT got completely lost (before going the the AAL wpt it was all good). The LEGS page is also not showing correct distances. The 29.2 Nm to AAL is correct (it later appeared that the USR waypoint was at the same position as AAL, I'm completely clueless what happened here)
![Screenshot (309)](https://user-images.githubusercontent.com/11409383/95659798-d594ed00-0b23-11eb-8ac0-31c75d90f7d8.png)

The PROGRESS page has its own opinion on the matter:
![Screenshot (310)](https://user-images.githubusercontent.com/11409383/95659839-273d7780-0b24-11eb-89a5-706102578a1d.png)

Here in the approach you see AAL and USR correspond. Where on earth does the 233 nm to AAL come from here??
![Screenshot (311)](https://user-images.githubusercontent.com/11409383/95659843-3290a300-0b24-11eb-8bdc-c65fa04272b6.png)

I really can't put my finger on this... It is completely inconsistent (at least when entering a flightplan on the ramp), that much we know...

**Comment by K20017** [comment]
The flight plan distances we are getting are computed by the Asobo flight plan manager and it is a horrible mess.  Sometimes it seems to work but then it will spit out numbers like you have seen and we have no idea where it's coming from.  We will investigate but we have plans to build our own flight plan manager to get around all the limitations and errors.

**Comment by TheFrett** [comment]
Ah that's a shame. I hope they fix this. It breaks the immersion when you get confused by three versions of enroute information. Hopefully a custom system as you say can overcome the default MSFS shortcomings. 

I will do some flights with the in-game planner tomorrow to see if I get the same results. I don't recall this mess from 0.3.1, then again we didn't have a proper LEGS page and FMS TEXT display so I guess it was there all along...

**Comment by J-Hoskin** [comment]
Alot of bugs you will encounter with this are indeed from Asobo's flight manager. Eg. It changes your active waypoint to previous around 5 miles before even reaching the active fix.

However, the FMS Text, the LEGS, and PROG page are calculating flight plan information differently to each other. They should all be the same wrong. So will we standardise this in coming releases.

Thanks for bring this up. Let us know any other similar problems :)

**Comment by TheFrett** [comment]
To add a bit more detail:

I just finished the flight (see also #232 based on the same flight) 

EKYT/26R VEPIP ADAXA LASMA EVLA/24

ILS24 Approach Transition LAROKSO (= via ROKSO fix)

Things go downhill right after passing the last enroute waypoint (LASMA in my case). From this point the situation is as follows:

- The FMS TEXT 'resets' to the initial flightplan and thus shows the first enroute waypoint as the next waypoint (red box in screenshot)
- The next waypoint indicated on the MFD (and PFD) is correct (ROKSO, green box in screenshot)
- On the LEGS page, a waypoint USER is inserted somehow, previous waypoint is RW26R (that is from my departure airport, this is probably related to the 'reset' of the FMS TEXT?). 

![Screenshot (316)](https://user-images.githubusercontent.com/11409383/95676046-a03bde00-0bbb-11eb-857a-c0959f067525.png)

**Comment by dga711** [comment]
I think I have it working for LEGS now. The USER waypoint is something the game generates when hitting approach (on some).
Need to fix approach things for FMS TEXT now.
